src/content_script/mastodonInject.js

```diff
@@ -0,0 +1,143 @@
+}
+
+/**
+ * Wait for element to appear.
+ * 
+ * @param {string} selector
+ * @param {boolean} multiple
+ * @param {number} timeoutDuration
+ * @see {@link https://github.com/storybookjs/test-runner/blob/6d41927154e8dd1e4c9e7493122e24e2739a7a0f/src/setup-page.ts#L134}
+ *  from which this was adapted
+ * @returns {Promise}
+ */
+function waitForElement(selector, multiple = false, timeoutDuration) {
+    return new Promise((resolve, reject) => {
+        const getElement = () => (
+            multiple
+                ? document.querySelectorAll(selector)
+                : document.querySelector(selector)
+        );
+        const isElementFound = (el) => (!multiple && el) || (multiple && el.length > 0);
+
+        const timeout = window.setTimeout(() => {
+            reject(new Error("waitForElement timed out"));
+        }, timeoutDuration);
+
+        const element = getElement();
+        if(isElementFound(element)){
+            window.clearTimeout(timeout);
+            return resolve(element);
+        }
+
+        const observer = new MutationObserver(() => {
+            const element = getElement();
+            if(isElementFound(element)){
+                window.clearTimeout(timeout);
+                resolve(element);
+                observer.disconnect();
+            }
+        });
+
+        observer.observe(document.body, {
+            childList: true,
+            subtree: true
+        });
+
+        return null;
+    }); 
+}
+
+/**
+ * Inject replacement onClick handler for Follow button.
+ * 
+ * @returns {void}
+ */
+async function injectFollowButton() {
+    try {
+        const followButton = await waitForElement(".account__header__tabs__buttons button:first-of-type", false, TIMEOUT_DURATION);
+        followButton.addEventListener("click", onClickFollow);
+    } catch (error) {
+        // Follow button failed to appear
+    }
+}
+
+/**
+ * Inject replacement onClick handler for Interaction buttons.
+ * 
+ * @returns {void}
+ */
+async function injectInteractionButtons() {
+    const INJECTED_REPLY_CLASS = "mastodon-simplified-federation-injected-interaction";
+    try {
+        const replyButtons = await waitForElement(
+            ".item-list[role='feed'] article[data-id] .status__action-bar button," +
+            ".detailed-status__wrapper .detailed-status__action-bar button",
+            true,
+            TIMEOUT_DURATION,
+        );
+        replyButtons.forEach((button) => {
+            if(!button.classList.contains(INJECTED_REPLY_CLASS)){
+                button.addEventListener("click", onClickInteract);
+                button.classList.add(INJECTED_REPLY_CLASS);
+            }
+        });
+    } catch (error) {
+        // Interaction buttons failed to appear
+        console.log(error);
+    }
+}
+
+/**
+ * Initialise injection for Mastodon Follow button.
+ * 
+ * @returns {void}
+ */
+function init() {
+    const observer = new MutationObserver(() => {
```

Comment:
Huh what is the reason for another MutationObserver here? I don't get it.

`waitForElement` already has one, does not it? So why need one here again, which alsonlives forever and is never cancelled.

Reply:
The MutationObserver is here to re-run the injection code when new posts are dynamically loaded via AJAX requests (as will happen when the user scrolls down). If we just run the injection code once, the interaction buttons in new posts won't be injected. 

I realise we could also catch the AJAX event using request interception in the background script, but that would require the background script to pass a message to the contentscript, which would make the code more complex. The MutationObserver approach also has the advantage of being more resilient (e.g. even if the API endpoint for loading more posts changes, the MutationObserver approach will continue to work).

Reply:
Ahh so can we limit that observer somewhat or only start it if we detect a mastokdn site,because
1. AFAIK this is a computation finally heavy operation
2. We inject that on all sites, so it must not break or should nor snow down these listening for unnecessary stuff

Reply:
And yeah I'm also against catching that request or so, let's keep it simple and just optimise this observer here.

Reply:
Got it. I'll use the `#mastodon` CSS selector to identify Mastodon instances.